### ISSUE TITLE
NEW StateManagment behavior

### ISSUE THREAD

**Issue by mikepenz** [mikepenz/FastAdapter#122]
I am currently trying and testing a NEW behavior of the FastAdapter which will allow more flexibility for developers. 

All changes are done in this branch:
https://github.com/mikepenz/FastAdapter/commits/feature/alternativeStateManagement

And a first demo application is also available:
https://drive.google.com/open?id=0B4Y8HN2BnSjpdG04MGk0cnMySmc

I would be happy If people could test this new behavior and report back. 

You can also test the whole thing in your app via this SNAPSHOT (available on the SNAPSHOT maven repo):
```gradle
compile('com.mikepenz:fastadapter:1.5.0.b2-SNAPSHOT@aar') {
	transitive = true
}

//extension:
compile 'com.mikepenz:fastadapter-extensions:1.5.0.b2-SNAPSHOT@aar'
//The tiny Materialize library used for its useful helper classes
compile 'com.mikepenz:materialize:0.8.8@aar'
```

**Comment by Rainer-Lang** [comment]
So selections are managed by the identifier?

**Comment by mikepenz** [comment]
@Rainer-Lang in the new behavior I have to go over the list to find selections and the expandable states, so the longer the list will get the slower this behavior will get. 

Wheras in the previous behavior I always kept the position of the selections and expandable states, so I was able to directly access the positions which are needed.

**Comment by artests87** [comment]
Is it  the specific thread? Tap again with no effect When I selected some items, I didn't deselect it.

**Comment by artests87** [comment]
Hi, Mike!
When I use method getPosition(Item item) from Adapter I always get "-1".
Maybe am I doing something wrong?
Thanks.

**Comment by mikepenz** [comment]
@artests87 yes ;). As the new `StateManagment` is just a pre release all discussion should go here, so we can pretest everything to work as expected. 

The first issue I can reproduce and will fix now. 

To the second. Does your item have an identifier?

**Comment by artests87** [comment]
And I don't have position subItems before I didn't click on ExpandableItem and SubItems will be show.

**Comment by artests87** [comment]
Yes. Al my items have ID

**Comment by artests87** [comment]
I don't select/deselect all (with subitems) . May will you create such methods?

**Comment by artests87** [comment]
And Invert method)

**Comment by artests87** [comment]
I'm trying to realise such methods myself, but all what I have looks like "без окон без дверей полна жопа огурцов".

**Comment by artests87** [comment]
And more one method. It have to looks like "open expandable item, without click, only programm's command"

**Comment by mikepenz** [comment]
@artests87 subItems do not have a position as long as they are not displayed. That's not possible as they are then currently not part of the list. 

Also you may group your answers. Otherwhise this thread will get a bit complicated :D

select / deselect - I will have a look, but this may be easier if you just go over the items, and notify the adapter that the items were changed. 

you can expand items already via the fastadapter

**Comment by artests87** [comment]
> you can expand items already via the fastadapter
toggleExpandable?

 Thanks a lot, Mike)

**Comment by mikepenz** [comment]
@artests87 I found the issue about deselection. Will be in the next SNAPSHOT. 

About expanding. I was talking about this method:
https://github.com/mikepenz/FastAdapter/blob/develop/library/src/main/java/com/mikepenz/fastadapter/FastAdapter.java#L1046

**Comment by mikepenz** [comment]
@artests87 new SNAPSHOT is available
```gradle
compile('com.mikepenz:fastadapter:1.5.0.b2-SNAPSHOT@aar') {
	transitive = true
}

//extension:
compile 'com.mikepenz:fastadapter-extensions:1.5.0.b2-SNAPSHOT@aar'
//The tiny Materialize library used for its useful helper classes
compile 'com.mikepenz:materialize:0.8.8@aar'
```

**Comment by artests87** [comment]
Deselect all, deselect on click are good now)
But I don't recognise, how to select all. When I go over the items, I don't have subitems' position before I expand Parent Item. And after that parents' position changed - my "go over" - turned into "game over".

**Comment by mikepenz** [comment]
@artests87 yes modified the `deselect()` method ;)
For this case I added a `select()` method for you :D

**Comment by artests87** [comment]
Thank you very much)
But now my select() selects all, but don't update Adapter. All changes I see only after "reopen" my adapter.
Because only withSelected(true) don't help us(
We need "select(position)" for everything)

**Comment by mikepenz** [comment]
@artests87 shouldn't be the case as I call: `notifyDataSetChanged();` after the select was done. I just tried it in the sample app and it worked, and instantly displayed all items selected.

**Comment by artests87** [comment]
Don't work.
Can you give me the code of the sample app?

**Comment by mikepenz** [comment]
https://github.com/mikepenz/FastAdapter/tree/develop/app


Just added `.select()` to the menu click in the sample app to try it out

**Comment by artests87** [comment]
Thanks a lot! Everything fine now.

**Comment by mikepenz** [comment]
@artests87 ok great. What was the issue?

**Comment by artests87** [comment]
I reconnected my adapter to the RecyclerView (rv.setAdapter(mActivity.getmFastItemAdapter());) before select(). But I didn't found when I lost the tie...

**Comment by artests87** [comment]
Hi, Mike!
Maybe you'll realise one more feature?
If I clicked "select all" - all items in my adapter selected, with ExpandableItems.
If I clicked "deselect all" - all items in my adapter deselected with ExpandableItems.
But if I clicked for each subItems (of one ExpandableItem), than deselect all of it, my parent ExpandableItems no deselect auto.
And if I clicked on first subItem - parent ExpandableItems no select auto.
Thanks.
